File: Kotlin/sort/QuickSort/src/QuickSort.kt
/*
QuickSort Algortihm works on Divide and Conquer Algorithm. It creates two empty arrays to hold elements less than the pivot value and
elements greater than the pivot value, and then recursively sort the sub arrays. There are two basic operations in the algorithm, 
swapping items in place and partitioning a section of the array.
*/

//Importing the Java.util package, it is needed because we are using the scanner object.
import java.util.*

//The function which will sort the given array in ascending order using QuickSort Algortihm
fun quicksort(items:List<Int>):List<Int>{
    //If there is only one element in the list, then there is no need to sort
    if (items.count() < 2){
        return items
    }
   
    //Vaiable pivot stores the index of middle element of the list
    val pivot = items[items.count()/2]
    
    //Variable equalto stores the elements at the pivot index
    val equalto = items.filter { it == pivot }
    
    //Variable lesser stores the list of element with indexes less than the pivot index
    val lesser = items.filter { it < pivot }
   
    //Variable greater stores the list of element with indexes less than the pivot index
    val greater = items.filter { it > pivot }
    
    //Calling the quicksort function recursively on the splitted arrays
    //This will get recursively called until left with multiple arrays with a single element and arranged in order
    return quicksort(lesser) + equalto + quicksort(greater)
}
//Main Driver Code
fun main() {
    val input = Scanner(System.`in`)

    println("Enter the length of the array")
    val arrayLength = input.nextInt()
    val arrayOfElements = arrayListOf<Int>()

    println("Enter the List of numbers you want to Sort:")
    for(index in 0 until arrayLength) {
        val element = input.nextInt()
        arrayOfElements.add(element)
    }
    print("Original List: ")
    for(index in 0..4)
    {
        val number: Int = arrayOfElements[index]
        print("$number\t")
    }
    print("\nOrdered list: ")
    val ordered =  quicksort(arrayOfElements)
    println(ordered)
}

/*
First Testcasee:
Enter the length of the array
6
Enter the List of numbers you want to Sort:
34
33
67
3
2
45
Original List: 34	33	67	3	2   45	
Ordered list: [2, 3, 33, 34, 45, 67]

Time Complexity: O(n log n)
Space Complexity: O(log n)
*/

File: PHP/sort/quick_sort.php
<!-- Quicksort is a divide-and-conquer algorithm.
    It works by selecting a 'pivot' element from the array
    and partitioning the other elements into two sub-arrays,
    according to whether they are less than or greater than the pivot. 
    The sub-arrays are then sorted recursively. -->

    <?php

function quickSort($arr)
 {
	$low = $great = array(); 
    /* If the array size is less than 2,
    we will return the array as it is */ 
	if(count($arr) < 2)
		return $arr;

    /*returns the key of the $arr element that's currently being pointed */
	$pivotKey = key($arr);
    /*removes the first element from $arr
      and stores the value in $pivot */
	$pivot = array_shift($arr);

	foreach($arr as $val)
        /* if the value is less than pivot update $low array */
		if($val <= $pivot)
			$low[] = $val;
        /*if the value is more than pivot update $great array*/
		elseif ($val > $pivot)
			$great[] = $val;

    /* merging the arrays to obtain the final result */
	return array_merge(quickSort($low),array($pivotKey=>$pivot),quickSort($great));
}

/*declaration and initialization of $arr*/
$arr = array(4, 0, 3, 1, -1, 6);
echo 'Original array is:    ';
$i=0;
for(;$i<count($arr)-1;$i++)
	echo $arr[$i].',';
echo $arr[$i].'<br>';

/*call of sort function*/
$arr = quickSort($arr);

/*printing the finally sorted array*/
echo 'Sorted array is:  ';
$i=0;
for(;$i<count($arr)-1;$i++)
	echo $arr[$i].',';
echo $arr[$i];

?>


<!-- Output Screen =>

Original array is: 4,0,3,1,-1,6
Sorted array is: -1,0,1,3,4,6

Time complexity : O(n^2)
Space complexity : O(log(n)) -->

